### PR TITLE
[Tabs] MDCTabBarViewController - like UITabBarViewController

### PR DESCRIPTION
**Pull request by DavidPhillipOster**
but uses Material Design's MDCTabBar

Usable. In future pull requests I'll add documentation, an Interface Builder example, and a Swift unit test.

cr/154488014

### CODE REVIEW

**Comment by DavidPhillipOster** [comment]
Interface Builder example added

File: components/Tabs/examples/TabBarViewControllerExample.m
Comment (ianegordon) [inline components/Tabs/examples/TabBarViewControllerExample.m]: nit: Confirm that this file is from 2016 and this is not a C&P error here and below.

File: components/Tabs/src/MDCTabBarViewController.m
Comment (ianegordon) [inline components/Tabs/src/MDCTabBarViewController.m]: Consider removing the #if DEBUG here.  NSAssert is disabled in RELEASE builds by default.

File: components/Tabs/src/MDCTabBarViewController.m
Comment (ianegordon) [inline components/Tabs/src/MDCTabBarViewController.m]: Add <UITabBarItem *> to array.

File: components/Tabs/src/MDCTabBarViewController.m
Comment (ianegordon) [inline components/Tabs/src/MDCTabBarViewController.m]: indenting

**Review by ianegordon — CHANGES_REQUESTED** [review]
Examples demonstrated in person.

File: components/Tabs/examples/TabBarViewControllerExample.m
Comment (DavidPhillipOster) [reply components/Tabs/examples/TabBarViewControllerExample.m]: These files are from 2017. Thanks for the correction

File: components/Tabs/src/MDCTabBarViewController.m
Comment (DavidPhillipOster) [reply components/Tabs/src/MDCTabBarViewController.m]: done

File: components/Tabs/src/MDCTabBarViewController.m
Comment (DavidPhillipOster) [reply components/Tabs/src/MDCTabBarViewController.m]: done

**Comment by ianegordon** [comment]
Local scripts/test_all succeeded.

**Review by ianegordon — APPROVED** [review]

**Comment by ianegordon** [comment]
Waiting on the +1 from Brian before Merge.

**Comment by brianjmoore** [comment]
Please run clang-format against all the new files here:
```
scripts/install_contributor_tools
clang-format -style=file -i FILE
```

File: components/Tabs/src/MDCTabBarViewController.h
Comment (brianjmoore) [inline components/Tabs/src/MDCTabBarViewController.h]: *view controllers

File: components/Tabs/src/MDCTabBarViewController.h
Comment (brianjmoore) [inline components/Tabs/src/MDCTabBarViewController.h]: *view controller

File: components/Tabs/src/MDCTabBarViewController.h
Comment (brianjmoore) [inline components/Tabs/src/MDCTabBarViewController.h]: nit here and below: first line of comment should be on its own line, like
```
/**
 The array of view controllers…
 */
```

File: components/Tabs/src/MDCTabBarViewController.h
Comment (brianjmoore) [inline components/Tabs/src/MDCTabBarViewController.h]: remove extra newline?

File: components/Tabs/src/MDCTabBarViewController.h
Comment (brianjmoore) [inline components/Tabs/src/MDCTabBarViewController.h]: omit "MDCTabBarViewController" here? I don't think it's needed

File: components/Tabs/src/MDCTabBarViewController.m
Comment (brianjmoore) [inline components/Tabs/src/MDCTabBarViewController.m:113]: Some brief comments breaking down the steps of this method might help readability a bit

File: components/Tabs/src/MDCTabBarViewController.m
Comment (brianjmoore) [inline components/Tabs/src/MDCTabBarViewController.m]: Does this fulfill the contract that returning NO prevents the tab switch? I think the tab bar will still switch and leave things inconsistent

Possibly relevant - there is a tabBar:willSelectItem: that would be a good hook for the willSelectViewController: part of this API, but no shouldSelectItem.

Comment (DavidPhillipOster) [reply components/Tabs/src/MDCTabBarViewController.m]: Modified  tabBar: to have shouldSelect, since it is a cover for UICollectionView's shouldSelect.

File: components/Tabs/src/private/MDCItemBar.m
Comment (DavidPhillipOster) [inline components/Tabs/src/private/MDCItemBar.m]: I ran this file through clang-format, but my intent was that the only significant change was from willSelectItem to shouldSelectItem.

Comment (brianjmoore) [reply components/Tabs/src/private/MDCItemBar.m]: Can you double check your setup? It looks like the style from material-components-ios/.clang-format isn't being applied - the mangling below seems to be mostly due to wrapping to 80 columns

File: components/Tabs/src/MDCTabBar.h
Comment (brianjmoore) [inline components/Tabs/src/MDCTabBar.h]: Unfortunately this is an API change that would affect behavior so we have to follow the correct process to ensure we don't break clients.

What this means here:
* We have to add the new method rather than replacing the old one and implement a backward-compatible behavior so existing clients are unaffected.
* The new method needs an API review. I think we can do this inline here.
* The old method should be deprecated following the documented process. See https://github.com/material-components/material-components-ios/blob/develop/contributing/deprecation_policy.md

File: components/Tabs/src/private/MDCItemBar.h
Comment (brianjmoore) [inline components/Tabs/src/private/MDCItemBar.h]: Since this is an internal, private interface let's be safe and keep everything required. You could actually delete willSelectItem from this protocol and just use shouldSelectItem to implement the MDCTabBar interface I think.

**Comment by brianjmoore** [comment]
Left some inline comments above.

**Review by ajsecord — CHANGES_REQUESTED** [review]
@ianegordon this seems to be missing the readme docs, the required files for the website generation, etc. Can we please land it into a feature branch until it's ready?

File: components/Tabs/src/private/MDCItemBar.h
Comment (DavidPhillipOster) [inline components/Tabs/src/private/MDCItemBar.h]: Done

File: components/Tabs/src/MDCTabBarViewController.h
Comment (DavidPhillipOster) [reply components/Tabs/src/MDCTabBarViewController.h]: Done

File: components/Tabs/src/private/MDCItemBar.h
Comment (DavidPhillipOster) [reply components/Tabs/src/private/MDCItemBar.h]: Done

File: components/Tabs/src/MDCTabBarViewController.m
Comment (DavidPhillipOster) [reply components/Tabs/src/MDCTabBarViewController.m:113]: I don't understand what you are asking for here.

File: components/Tabs/src/MDCTabBar.h
Comment (brianjmoore) [inline components/Tabs/src/MDCTabBar.h]: Nit: "NOTE" instead of "TODO"

File: components/Tabs/src/MDCTabBar.m
Comment (brianjmoore) [inline components/Tabs/src/MDCTabBar.m]: -2 spaces (clang-format should take care of this)

File: components/Tabs/src/MDCTabBarViewController.h
Comment (brianjmoore) [inline components/Tabs/src/MDCTabBarViewController.h]: remove extra newline?

File: components/Tabs/src/MDCTabBarViewController.m
Comment (brianjmoore) [reply components/Tabs/src/MDCTabBarViewController.m:113]: What I meant was that it takes a bit too much close reading to understand the high-level steps of this method: updating the property and showing the newly-visible view controller. (Similarly for the method above where it's not obvious that the first chunk of code is detaching removed child view controllers and the rest is updating the property). 

It's just about skimmability - Two brief comments separating out and denoting those two responsibilities would help future readers skim and understand the code more quickly.

File: components/Tabs/src/private/MDCItemBar.m
Comment (brianjmoore) [inline components/Tabs/src/private/MDCItemBar.m:431]: Yeah, agreed that clang-format is wrong here. "//clang-format on/off" could help preserve the niceness or we could just accept the slight readability hit and welcome our new auto-formatting overlords.

**Review by brianjmoore — APPROVED** [review]
Approved, pending a few inline comments and Adrian's request above. Thank you!

**Comment by brianjmoore** [comment]
FYI: Bottom navigation support has landed, so you should be good to adopt it here. Should just need to implement -[UIBarPositioningDelegate positionForBar:] and return UIBarPositionBottom.

File: components/Tabs/src/MDCTabBarViewController.m
Comment (DavidPhillipOster) [reply components/Tabs/src/MDCTabBarViewController.m:113]: Done

File: components/Tabs/src/private/MDCItemBar.m
Comment (DavidPhillipOster) [reply components/Tabs/src/private/MDCItemBar.m:431]: done

**Comment by DavidPhillipOster** [comment]
• I believe I've addressed all of Brianjmoore's issues at this point. 
• Ajsecord asked for a readme file. I could use some help with that. 
• He also asked for 'the required files for the website generation' - I don't understand that at all - what specifically is missing?

**Comment by ianegordon** [comment]
Note this PR got automagically close as part of the switching base branches.

It has landed in feature-tabcontroller.

**Pull request closed without merging**
